File: minimax_algorithm.py
# This work is heavily influenced by the following project:
# https://github.com/colingogogo/gobang_AI

class MinimaxAlgorithm:
    def __init__(self, board_size=(12, 12), search_depth=3, attack_ratio=1):
        """
        Initialize the Minimax algorithm
        :param board_size: Board size, default 15x15
        :param search_depth: Search depth, default 3 (must be odd)
        :param attack_ratio: Attack ratio (>1 is more aggressive, <1 is more defensive)
        """
        self.COLUMN, self.ROW = board_size
        self.DEPTH = search_depth
        self.ratio = attack_ratio
        
        # Statistics data
        self.cut_count = 0    # Pruning count
        self.search_count = 0 # Search count
        
        # Board data
        self.player_pieces = []   # AI pieces
        self.opponent_pieces = [] # Opponent pieces
        self.all_pieces = []      # All pieces
        self.all_positions = []   # All board positions
        self.next_move = [0, 0]   # Position AI will place next
        
        # Initialize all board positions
        for i in range(self.COLUMN+1):
            for j in range(self.ROW+1):
                self.all_positions.append((i, j))
        
        # Shape scoring table
        self.shape_score = [
            (50, (0, 1, 1, 0, 0)),
            (50, (0, 0, 1, 1, 0)),
            (200, (1, 1, 0, 1, 0)),
            (500, (0, 0, 1, 1, 1)),
            (500, (1, 1, 1, 0, 0)),
            (5000, (0, 1, 1, 1, 0)),
            (5000, (0, 1, 0, 1, 1, 0)),
            (5000, (0, 1, 1, 0, 1, 0)),
            (5000, (1, 1, 1, 0, 1)),
            (5000, (1, 1, 0, 1, 1)),
            (5000, (1, 0, 1, 1, 1)),
            (5000, (1, 1, 1, 1, 0)),
            (5000, (0, 1, 1, 1, 1)),
            (50000, (0, 1, 1, 1, 1, 0)),
            (99999999, (1, 1, 1, 1, 1))
        ]
    
    def get_next_move(self, player_pieces, opponent_pieces):
        """
        Get the best placement position for the AI's next move
        :param player_pieces: List of AI's placed pieces, each element is (x,y) coordinate
        :param opponent_pieces: List of opponent's placed pieces
        :return: (x,y) tuple representing the best placement position
        """
        self.player_pieces = player_pieces.copy()
        self.opponent_pieces = opponent_pieces.copy()
        self.all_pieces = player_pieces.copy() + opponent_pieces.copy()
        
        # Reset statistics
        self.cut_count = 0
        self.search_count = 0
        
        # Run the Minimax algorithm
        self.negamax(True, self.DEPTH, -99999999, 99999999)
        
        # Return the best move
        return (self.next_move[0], self.next_move[1])
    
    def get_statistics(self):
        """Get algorithm search statistics"""
        return {
            "cut_count": self.cut_count,
            "search_count": self.search_count
        }
    
    def negamax(self, is_ai, depth, alpha, beta):
        """
        Negamax search algorithm with alpha-beta pruning
        :param is_ai: Whether it is the AI's turn
        :param depth: Current search depth
        :param alpha: Alpha value
        :param beta: Beta value
        :return: Optimal score for the current position
        """
        # Check if the game is over or if the search depth is reached
        if self.check_win(self.player_pieces) or self.check_win(self.opponent_pieces) or depth == 0:
            return self.evaluation(is_ai)
        
        # Get all empty positions
        blank_list = list(set(self.all_positions).difference(set(self.all_pieces)))
        # Sort search order to improve pruning efficiency
        self.order_moves(blank_list)
        
        # Iterate through each candidate move
        for next_step in blank_list:
            self.search_count += 1
            
            # Skip positions without adjacent pieces (reduce computation)
            if not self.has_neighbor(next_step):
                continue
            
            # Simulate placing a piece
            if is_ai:
                self.player_pieces.append(next_step)
            else:
                self.opponent_pieces.append(next_step)
            self.all_pieces.append(next_step)
            
            # Recursive search
            value = -self.negamax(not is_ai, depth - 1, -beta, -alpha)
            
            # Undo the move
            if is_ai:
                self.player_pieces.remove(next_step)
            else:
                self.opponent_pieces.remove(next_step)
            self.all_pieces.remove(next_step)
            
            # Update the best value
            if value > alpha:
                if depth == self.DEPTH:
                    self.next_move[0] = next_step[0]
                    self.next_move[1] = next_step[1]
                
                # Alpha-beta pruning
                if value >= beta:
                    self.cut_count += 1
                    return beta
                alpha = value
        
        return alpha
    
    def order_moves(self, blank_list):
        """
        Sort possible placement positions, prioritizing those close to the last placed position
        :param blank_list: List of empty positions
        """
        if not self.all_pieces:
            return
            
        last_pt = self.all_pieces[-1]
        for i in range(-1, 2):
            for j in range(-1, 2):
                if i == 0 and j == 0:
                    continue
                    
                pos = (last_pt[0] + i, last_pt[1] + j)
                if pos in blank_list:
                    blank_list.remove(pos)
                    blank_list.insert(0, pos)
    
    def has_neighbor(self, point):
        """
        Check if a position has adjacent pieces
        :param point: Position coordinate
        :return: Boolean value
        """
        for i in range(-1, 2):
            for j in range(-1, 2):
                if i == 0 and j == 0:
                    continue
                    
                neighbor = (point[0] + i, point[1] + j)
                if neighbor in self.all_pieces:
                    return True
        return False
    
    def evaluation(self, is_ai):
        """
        Evaluate the score of the current position
        :param is_ai: Whether to evaluate from the AI's perspective
        :return: Score value
        """
        if is_ai:
            my_list = self.player_pieces
            enemy_list = self.opponent_pieces
        else:
            my_list = self.opponent_pieces
            enemy_list = self.player_pieces
        
        # Calculate the score for oneself
        score_all_arr = []
        my_score = 0
        for pt in my_list:
            m, n = pt
            my_score += self.cal_score(m, n, 0, 1, enemy_list, my_list, score_all_arr)
            my_score += self.cal_score(m, n, 1, 0, enemy_list, my_list, score_all_arr)
            my_score += self.cal_score(m, n, 1, 1, enemy_list, my_list, score_all_arr)
            my_score += self.cal_score(m, n, -1, 1, enemy_list, my_list, score_all_arr)
        
        # Calculate the score for the enemy
        score_all_arr_enemy = []
        enemy_score = 0
        for pt in enemy_list:
            m, n = pt
            enemy_score += self.cal_score(m, n, 0, 1, my_list, enemy_list, score_all_arr_enemy)
            enemy_score += self.cal_score(m, n, 1, 0, my_list, enemy_list, score_all_arr_enemy)
            enemy_score += self.cal_score(m, n, 1, 1, my_list, enemy_list, score_all_arr_enemy)
            enemy_score += self.cal_score(m, n, -1, 1, my_list, enemy_list, score_all_arr_enemy)
        
        # Total score = My score - Enemy score * ratio * 0.1
        return my_score - enemy_score * self.ratio * 0.1
    
    def cal_score(self, m, n, x_direct, y_direct, enemy_list, my_list, score_all_arr):
        """
        Calculate the score in a specific direction
        :param m, n: Piece position
        :param x_direct, y_direct: Direction vector
        :param enemy_list: Enemy pieces
        :param my_list: My pieces
        :param score_all_arr: Calculated score shapes
        :return: Score in that direction
        """
        add_score = 0
        max_score_shape = (0, None)
        
        # Check if this direction has been calculated
        for item in score_all_arr:
            for pt in item[1]:
                if m == pt[0] and n == pt[1] and x_direct == item[2][0] and y_direct == item[2][1]:
                    return 0
        
        # Scan in a specific direction to find shapes
        for offset in range(-5, 1):
            pos = []
            for i in range(6):
                point = (m + (i + offset) * x_direct, n + (i + offset) * y_direct)
                if point in enemy_list:
                    pos.append(2)
                elif point in my_list:
                    pos.append(1)
                else:
                    pos.append(0)
            
            tmp_shape5 = tuple(pos[:5])
            tmp_shape6 = tuple(pos)
            
            # Match shapes and score
            for score, shape in self.shape_score:
                if tmp_shape5 == shape or tmp_shape6 == shape:
                    if score > max_score_shape[0]:
                        max_score_shape = (score, tuple(
                            (m + (i + offset) * x_direct, n + (i + offset) * y_direct)
                            for i in range(5)
                        ), (x_direct, y_direct))
        
        # Calculate cross-score for shapes
        if max_score_shape[1] is not None:
            for item in score_all_arr:
                for pt1 in item[1]:
                    for pt2 in max_score_shape[1]:
                        if pt1 == pt2 and max_score_shape[0] > 10 and item[0] > 10:
                            add_score += item[0] + max_score_shape[0]
            
            score_all_arr.append(max_score_shape)
        
        return add_score + max_score_shape[0]
    
    def check_win(self, pieces):
        """
        Check for victory
        :param pieces: List of pieces
        :return: Boolean value
        """
        for m in range(self.COLUMN):
            for n in range(self.ROW):
                # Check horizontal direction
                if (n < self.ROW - 4 and 
                    (m, n) in pieces and 
                    (m, n + 1) in pieces and 
                    (m, n + 2) in pieces and 
                    (m, n + 3) in pieces and 
                    (m, n + 4) in pieces):
                    return True
                    
                # Check vertical direction
                elif (m < self.COLUMN - 4 and 
                      (m, n) in pieces and 
                      (m + 1, n) in pieces and 
                      (m + 2, n) in pieces and 
                      (m + 3, n) in pieces and 
                      (m + 4, n) in pieces):
                    return True
                    
                # Check right diagonal
                elif (m < self.COLUMN - 4 and n < self.ROW - 4 and
                      (m, n) in pieces and 
                      (m + 1, n + 1) in pieces and 
                      (m + 2, n + 2) in pieces and 
                      (m + 3, n + 3) in pieces and 
                      (m + 4, n + 4) in pieces):
                    return True
                    
                # Check left diagonal
                elif (m < self.COLUMN - 4 and n > 3 and
                      (m, n) in pieces and 
                      (m + 1, n - 1) in pieces and 
                      (m + 2, n - 2) in pieces and 
                      (m + 3, n - 3) in pieces and 
                      (m + 4, n - 4) in pieces):
                    return True
                    
        return False

File: minimax_algorithm.cpp
#include "minimax_algorithm.h"
#include <iostream>
#include <algorithm>

// Constructor implementation
MinimaxAlgorithm::MinimaxAlgorithm(std::pair<int, int> board_size, int search_depth, double attack_ratio) {
    // Initialize basic parameters
    COLUMN = board_size.first;
    ROW = board_size.second;
    DEPTH = search_depth;
    ratio = attack_ratio;
    
    // Initialize statistics
    cut_count = 0;
    search_count = 0;
    
    // Initialize next move
    next_move = {0, 0};
    
    // Initialize all possible board positions
    for (int i = 0; i <= COLUMN; i++) {
        for (int j = 0; j <= ROW; j++) {
            all_positions.push_back({i, j});
        }
    }
    
    // Initialize shape scoring table
    shape_score = {
        {50, {0, 1, 1, 0, 0}},
        {50, {0, 0, 1, 1, 0}},
        {200, {1, 1, 0, 1, 0}},
        {500, {0, 0, 1, 1, 1}},
        {500, {1, 1, 1, 0, 0}},
        {5000, {0, 1, 1, 1, 0}},
        {5000, {0, 1, 0, 1, 1, 0}},
        {5000, {0, 1, 1, 0, 1, 0}},
        {5000, {1, 1, 1, 0, 1}},
        {5000, {1, 1, 0, 1, 1}},
        {5000, {1, 0, 1, 1, 1}},
        {5000, {1, 1, 1, 1, 0}},
        {5000, {0, 1, 1, 1, 1}},
        {50000, {0, 1, 1, 1, 1, 0}},
        {99999999, {1, 1, 1, 1, 1}}
    };
}

std::pair<int, int> MinimaxAlgorithm::get_next_move(
    const std::vector<std::pair<int, int>>& player_pieces_input, 
    const std::vector<std::pair<int, int>>& opponent_pieces_input
) {
    // Copy the input pieces
    player_pieces = player_pieces_input;
    opponent_pieces = opponent_pieces_input;
    
    // Create all_pieces by combining player and opponent pieces
    all_pieces = player_pieces;
    all_pieces.insert(all_pieces.end(), opponent_pieces.begin(), opponent_pieces.end());
    
    // Reset statistics
    cut_count = 0;
    search_count = 0;
    
    // Run the Minimax algorithm
    negamax(true, DEPTH, -99999999, 99999999);
    
    // Return the best move
    return next_move;
}

std::map<std::string, int> MinimaxAlgorithm::get_statistics() const {
    return {
        {"cut_count", cut_count},
        {"search_count", search_count}
    };
}

int MinimaxAlgorithm::negamax(bool is_ai, int depth, int alpha, int beta) {
    // Check if the game is over or if the search depth is reached
    if (check_win(player_pieces) || check_win(opponent_pieces) || depth == 0) {
        return evaluation(is_ai);
    }
    
    // Get all empty positions by finding differences between all positions and placed pieces
    std::vector<std::pair<int, int>> blank_list;
    for (const auto& pos : all_positions) {
        if (std::find(all_pieces.begin(), all_pieces.end(), pos) == all_pieces.end()) {
            blank_list.push_back(pos);
        }
    }
    
    // Sort search order to improve pruning efficiency
    order_moves(blank_list);
    
    // Iterate through each candidate move
    for (const auto& next_step : blank_list) {
        search_count++;
        
        // Skip positions without adjacent pieces (reduce computation)
        if (!has_neighbor(next_step)) {
            continue;
        }
        
        // Simulate placing a piece
        if (is_ai) {
            player_pieces.push_back(next_step);
        } else {
            opponent_pieces.push_back(next_step);
        }
        all_pieces.push_back(next_step);
        
        // Recursive search
        int value = -negamax(!is_ai, depth - 1, -beta, -alpha);
        
        // Undo the move
        if (is_ai) {
            player_pieces.pop_back();
        } else {
            opponent_pieces.pop_back();
        }
        all_pieces.pop_back();
        
        // Update the best value
        if (value > alpha) {
            if (depth == DEPTH) {
                next_move = next_step;
            }
            
            // Alpha-beta pruning
            if (value >= beta) {
                cut_count++;
                return beta;
            }
            alpha = value;
        }
    }
    
    return alpha;
}

void MinimaxAlgorithm::order_moves(std::vector<std::pair<int, int>>& blank_list) {
    if (all_pieces.empty()) {
        return;
    }
    
    auto last_pt = all_pieces.back();
    
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            if (i == 0 && j == 0) {
                continue;
            }
            
            std::pair<int, int> pos = {last_pt.first + i, last_pt.second + j};
            auto it = std::find(blank_list.begin(), blank_list.end(), pos);
            
            if (it != blank_list.end()) {
                blank_list.erase(it);
                blank_list.insert(blank_list.begin(), pos);
            }
        }
    }
}

bool MinimaxAlgorithm::has_neighbor(const std::pair<int, int>& point) {
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            if (i == 0 && j == 0) {
                continue;
            }
            
            std::pair<int, int> neighbor = {point.first + i, point.second + j};
            if (std::find(all_pieces.begin(), all_pieces.end(), neighbor) != all_pieces.end()) {
                return true;
            }
        }
    }
    return false;
}

int MinimaxAlgorithm::evaluation(bool is_ai) {
    std::vector<std::pair<int, int>> my_list;
    std::vector<std::pair<int, int>> enemy_list;
    
    if (is_ai) {
        my_list = player_pieces;
        enemy_list = opponent_pieces;
    } else {
        my_list = opponent_pieces;
        enemy_list = player_pieces;
    }
    
    // Calculate the score for oneself
    std::vector<std::tuple<int, std::vector<std::pair<int, int>>, std::pair<int, int>>> score_all_arr;
    int my_score = 0;
    
    for (const auto& pt : my_list) {
        int m = pt.first;
        int n = pt.second;
        my_score += cal_score(m, n, 0, 1, enemy_list, my_list, score_all_arr);
        my_score += cal_score(m, n, 1, 0, enemy_list, my_list, score_all_arr);
        my_score += cal_score(m, n, 1, 1, enemy_list, my_list, score_all_arr);
        my_score += cal_score(m, n, -1, 1, enemy_list, my_list, score_all_arr);
    }
    
    // Calculate the score for the enemy
    std::vector<std::tuple<int, std::vector<std::pair<int, int>>, std::pair<int, int>>> score_all_arr_enemy;
    int enemy_score = 0;
    
    for (const auto& pt : enemy_list) {
        int m = pt.first;
        int n = pt.second;
        enemy_score += cal_score(m, n, 0, 1, my_list, enemy_list, score_all_arr_enemy);
        enemy_score += cal_score(m, n, 1, 0, my_list, enemy_list, score_all_arr_enemy);
        enemy_score += cal_score(m, n, 1, 1, my_list, enemy_list, score_all_arr_enemy);
        enemy_score += cal_score(m, n, -1, 1, my_list, enemy_list, score_all_arr_enemy);
    }
    
    // Total score = My score - Enemy score * ratio * 0.1
    return my_score - static_cast<int>(enemy_score * ratio * 0.1);
}

int MinimaxAlgorithm::cal_score(
    int m, int n, int x_direct, int y_direct, 
    const std::vector<std::pair<int, int>>& enemy_list,
    const std::vector<std::pair<int, int>>& my_list,
    std::vector<std::tuple<int, std::vector<std::pair<int, int>>, std::pair<int, int>>>& score_all_arr
) {
    int add_score = 0;
    std::pair<int, std::vector<std::pair<int, int>>> max_score_shape = {0, {}};
    std::pair<int, int> direction = {x_direct, y_direct};
    
    // Check if this direction has been calculated
    for (const auto& item : score_all_arr) {
        for (const auto& pt : std::get<1>(item)) {
            if (m == pt.first && n == pt.second && 
                x_direct == std::get<2>(item).first && 
                y_direct == std::get<2>(item).second) {
                return 0;
            }
        }
    }
    
    // Scan in a specific direction to find shapes
    for (int offset = -5; offset < 1; offset++) {
        std::vector<int> pos;
        std::vector<std::pair<int, int>> positions;
        
        for (int i = 0; i < 6; i++) {
            std::pair<int, int> point = {m + (i + offset) * x_direct, n + (i + offset) * y_direct};
            positions.push_back(point);
            
            if (std::find(enemy_list.begin(), enemy_list.end(), point) != enemy_list.end()) {
                pos.push_back(2);
            } else if (std::find(my_list.begin(), my_list.end(), point) != my_list.end()) {
                pos.push_back(1);
            } else {
                pos.push_back(0);
            }
        }
        
        std::vector<int> tmp_shape5(pos.begin(), pos.begin() + 5);
        
        // Match shapes and score
        for (const auto& shape_pair : shape_score) {
            int score = shape_pair.first;
            const auto& shape = shape_pair.second;
            
            bool matched = false;
            
            // Check for 5-length pattern
            if (shape.size() == 5) {
                matched = std::equal(shape.begin(), shape.end(), tmp_shape5.begin());
            } 
            // Check for 6-length pattern
            else if (shape.size() == 6) {
                matched = std::equal(shape.begin(), shape.end(), pos.begin());
            }
            
            if (matched && score > max_score_shape.first) {
                std::vector<std::pair<int, int>> shape_positions;
                for (int i = 0; i < 5; i++) {
                    shape_positions.push_back({m + (i + offset) * x_direct, n + (i + offset) * y_direct});
                }
                max_score_shape = {score, shape_positions};
            }
        }
    }
    
    // Calculate cross-score for shapes
    if (!max_score_shape.second.empty()) {
        for (const auto& item : score_all_arr) {
            for (const auto& pt1 : std::get<1>(item)) {
                for (const auto& pt2 : max_score_shape.second) {
                    if (pt1 == pt2 && max_score_shape.first > 10 && std::get<0>(item) > 10) {
                        add_score += std::get<0>(item) + max_score_shape.first;
                    }
                }
            }
        }
        
        score_all_arr.push_back({max_score_shape.first, max_score_shape.second, direction});
    }
    
    return add_score + max_score_shape.first;
}

bool MinimaxAlgorithm::check_win(const std::vector<std::pair<int, int>>& pieces) {
    for (int m = 0; m < COLUMN; m++) {
        for (int n = 0; n < ROW; n++) {
            // Check horizontal direction
            if (n < ROW - 4 &&
                std::find(pieces.begin(), pieces.end(), std::make_pair(m, n)) != pieces.end() &&
                std::find(pieces.begin(), pieces.end(), std::make_pair(m, n + 1)) != pieces.end() &&
                std::find(pieces.begin(), pieces.end(), std::make_pair(m, n + 2)) != pieces.end() &&
                std::find(pieces.begin(), pieces.end(), std::make_pair(m, n + 3)) != pieces.end() &&
                std::find(pieces.begin(), pieces.end(), std::make_pair(m, n + 4)) != pieces.end()) {
                return true;
            }
            
            // Check vertical direction
            else if (m < COLUMN - 4 &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 1, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 2, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 3, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 4, n)) != pieces.end()) {
                return true;
            }
            
            // Check right diagonal
            else if (m < COLUMN - 4 && n < ROW - 4 &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 1, n + 1)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 2, n + 2)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 3, n + 3)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 4, n + 4)) != pieces.end()) {
                return true;
            }
            
            // Check left diagonal
            else if (m < COLUMN - 4 && n > 3 &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m, n)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 1, n - 1)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 2, n - 2)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 3, n - 3)) != pieces.end() &&
                     std::find(pieces.begin(), pieces.end(), std::make_pair(m + 4, n - 4)) != pieces.end()) {
                return true;
            }
        }
    }
    
    return false;
} 